File: books/mazes-in-python/binary_tree.py
import random
from grid import Grid

def binary_tree(grid: Grid):
    for cell in grid:
        neighbor = random.choice([cell.north, cell.east])
        if neighbor:
            cell.link(neighbor)


if __name__ == "__main__":
    g = Grid(5, 5)
    binary_tree(g)

File: books/mazes-in-go/binary_tree.go
package mazes

import "math/rand"

func BinaryTree(grid *Grid) {
	grid.EachCells(func(cell *Cell) {
		if neighbor := chooseNeighbor(grid, cell); neighbor != nil {
			cell.link(neighbor)
		}
	})
}

func chooseNeighbor(grid *Grid, cell *Cell) *Cell {
	north := grid.North(cell)
	east := grid.East(cell)

	if north != nil && east != nil {
		if rand.Intn(2) == 0 {
			return north
		}
		return east
	}
	if north != nil {
		return north
	}
	if east != nil {
		return east
	}
	return nil
}
